### PR TITLE
[ZEPPELIN-1371]add text/numeric conversion support to table display

### PR DESCRIPTION
**Pull request by Peilin-Yang**
### What is this PR for?

We people in Twitter have great demand of adding a flexible drop down menu for the columns in the tables which enables the text/numeric conversion.
This is because people want some columns to be of string type which, for example, fits to the underlying DB definition.
The use cases of the change include:
1. For now sorting on the columns is always lexicographically because Zeppelin front end treats the data as strings. It the values in the table can be convertible between string and number then we can sort the column by numeric values. I have filed another ticket for this: https://issues.apache.org/jira/browse/ZEPPELIN-1372
2. Since the users know more about their data than us it would be nice if we let the users decide what is the real type of their data. It is annoying if user wants the column to be strings but the front end forcefully inserts commas in it. In some scenarios, users may also want to copy/paste the table to somewhere else. If people want to remove the commas before other actions then that will be a nightmare.... 
### What type of PR is it?

[Improvement | Feature]
### Todos
- [ ] - Task
### What is the Jira issue?
- https://issues.apache.org/jira/browse/ZEPPELIN-1371
### How should this be tested?
1. Click on the dropdown menu would convert the text/number of that column.
2. Other functionalities esp. the sorting function should not be affected.
### Screenshots (if appropriate)

image
image
### Questions:
- Does the licenses files need update? No
- Is there breaking changes for older versions? Probably a noticeable change in UI
- Does this needs documentation? No

### CODE REVIEW

**Comment by Peilin-Yang** [comment]
@corneadoug It was really messy to rebase so I created this new PR. The previous one was closed.

**Comment by corneadoug** [comment]
@Peilin-Yang Selenium failure should now be fixed on Master, can you rebase?

**Comment by Peilin-Yang** [comment]
@corneadoug done

**Comment by corneadoug** [comment]
@Peilin-Yang 
Have to fix a test, need to change the expected result in that test from `1,787` to `1787`
```
Failed tests: 
  SparkParagraphIT.testSqlSpark:188 Paragraph from SparkParagraphIT of testSqlSpark result: 
Expected: "age\njob\nmarital\neducation\nbalance\n30 unemployed married primary 1,787"
     but: was "age\nâ–¼\njob\nâ–¼\nmarital\nâ–¼\neducation\nâ–¼\nbalance\nâ–¼\n30 unemployed married primary 1787"

Tests run: 15, Failures: 1, Errors: 0, Skipped: 0
```

**Comment by Peilin-Yang** [comment]
@corneadoug I have changed the test case. Could you please review?

**Comment by corneadoug** [comment]
@Peilin-Yang The button we introduced in this PR, is breaking another test.
Although the test looks funny, let's fix the expected result too

```
Failed tests: 
  SparkParagraphIT.testSqlSpark:188 Paragraph from SparkParagraphIT of testSqlSpark result: 
Expected: "age\njob\nmarital\neducation\nbalance\n30 unemployed married primary 1787"
     but: was "age\nâ–¼\njob\nâ–¼\nmarital\nâ–¼\neducation\nâ–¼\nbalance\nâ–¼\n30 unemployed married primary 1787"
```

**Comment by Peilin-Yang** [comment]
@corneadoug It was very weird.
Can you tell how to test?
I have changed the numeric value 1,787 to 1787 but it seems that the failure is because of (maybe) some encoding/decoding at the first column.

**Comment by corneadoug** [comment]
Not really, there is no issue with the feature, the test itself is a bit weird because it transform to text whatever it find:
https://github.com/apache/zeppelin/blob/b8755ebb25ad793daa6873acc2e00b8d69588188/zeppelin-server/src/test/java/org/apache/zeppelin/integration/SparkParagraphIT.java#L189

Just need to change the expected value by adding that: `\nâ–¼` instead of `\n` in between the header column I guess

**Comment by Peilin-Yang** [comment]
@corneadoug updated

**Comment by corneadoug** [comment]
@Peilin-Yang Sorry that it takes so long, you will need to rebase

**Comment by corneadoug** [comment]
Also there is still the selenium test error:
```
Failed tests: 
  SparkParagraphIT.testSqlSpark:188 Paragraph from SparkParagraphIT of testSqlSpark result: 
Expected: "age\nÃ¢â€“Â¼\njob\nÃ¢â€“Â¼\nmarital\nÃ¢â€“Â¼\neducation\nÃ¢â€“Â¼\nbalance\nÃ¢â€“Â¼\n30 unemployed married primary 1787"
     but: was "age\nâ–¼\njob\nâ–¼\nmarital\nâ–¼\neducation\nâ–¼\nbalance\nâ–¼\n30 unemployed married primary 1787"
```

**Comment by Peilin-Yang** [comment]
@corneadoug can you see whether it can pass the test now?

**Comment by corneadoug** [comment]
Current failure is really weird, can you try launching the CI again (close/re-open PR)?

**Comment by Peilin-Yang** [comment]
@corneadoug do you have chance to look at what is wrong with the commit? thx

**Comment by corneadoug** [comment]
@Peilin-Yang Sorry I just updated to Sierra and am having a few issues.
From the log it seems there is some broken front-end tests (probably due to a rebase)
You should be able to run it locally in zeppelin-web folder with `./grunt test`

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 1]
```diff
@@ -30,12 +30,16 @@
     'websocketMsgSrv',
     'baseUrlSrv',
     'ngToast',
-    'saveAsService'
+    'saveAsService',
+    'esriLoader',
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: Removing esriLoader should fix the CI

File: zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js [hunk 2]
```diff
@@ -30,12 +30,16 @@
...
+    'handsonHelperService',
+    'dataTypeService'
   ];
 
   function ParagraphCtrl($scope, $rootScope, $route, $window, $routeParams, $location,
                          $timeout, $compile, $http, $q, websocketMsgSrv,
-                         baseUrlSrv, ngToast, saveAsService) {
+                         baseUrlSrv, ngToast, saveAsService, esriLoader,
```
Comment (corneadoug) [inline zeppelin-web/src/app/notebook/paragraph/paragraph.controller.js]: Removing esriLoader here too, should fix the CI

**Comment by Peilin-Yang** [comment]
@corneadoug I've tested using `./grunt test` and it passed the test. But CI check still failed.

**Comment by corneadoug** [comment]
@Peilin-Yang still the same test failing but asking for different value this time...
```
Failed tests: 
  SparkParagraphIT.testSqlSpark:188 Paragraph from SparkParagraphIT of testSqlSpark result: 
Expected: "age\nÃƒÂ¢Ã¢â‚¬â€œÃ‚Â¼\njob\nÃƒÂ¢Ã¢â‚¬â€œÃ‚Â¼\nmarital\nÃƒÂ¢Ã¢â‚¬â€œÃ‚Â¼\neducation\nÃƒÂ¢Ã¢â‚¬â€œÃ‚Â¼\nbalance\nÃƒÂ¢Ã¢â‚¬â€œÃ‚Â¼\n30 unemployed married primary 1787"
     but: was "age\nâ–¼\njob\nâ–¼\nmarital\nâ–¼\neducation\nâ–¼\nbalance\nâ–¼\n30 unemployed married primary 1787"

Tests run: 16, Failures: 1, Errors: 0, Skipped: 0
```

I will see how we can change this test

**Comment by Peilin-Yang** [comment]
@corneadoug thanks, the error is really strange.

**Comment by bzz** [comment]
@Peilin-Yang I think it's very useful feature - do you think you will have some time to rebase and see if that helps? Just FIY - frontend tests got moved under `npm run test`

**Comment by Peilin-Yang** [comment]
@bzz done

**Comment by Peilin-Yang** [comment]
@bzz able to merge?

**Comment by bzz** [comment]
@Peilin-Yang CI is green now, awesome, thank you! 
Sorry for sloppy reply, let me try it.

\cc @felizbear for a quick review as well

**Comment by felizbear** [comment]
@Peilin-Yang a quick question: for some reason, when setting `value` to date type i get this

screen shot 2016-11-30 at 2 02 08 pm

it looks somewhat inconsistent to me, but maybe i don't understand something

File: zeppelin-web/src/app/handsontable/handsonHelper.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  } else if (value.length > '%html'.length && '%html ' === value.substring(0, '%html '.length)) {
+    td.innerHTML = value.substring('%html'.length);
+  } else {
+    Handsontable.renderers.TextRenderer.apply(this, arguments);
+  }
+}
+
+function _dateValidator(value, callback) {
+  var d = moment(value);
+  return callback(d.isValid() ? true : false);
```
Comment (felizbear) [inline zeppelin-web/src/app/handsontable/handsonHelper.js]: do you mind changing this to `return callback(d.isValid());`?

`d.isValid()` already returns boolean, so the ternary creates unnecessary cognitive strain for the reader

Comment (Peilin-Yang) [inline zeppelin-web/src/app/handsontable/handsonHelper.js]: okay

**Comment by Peilin-Yang** [comment]
@felizbear 
The code of this PR uses `momentjs` to parse the string to date.
Specifically, http://momentjs.com/docs/#/parsing/string/ 
Integers are not of ISO8601 format so javascript's native method `Date` is called to parse the string.
More details can be found at  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse
In your example, integer strings smaller than 13 are parsed as `nn/01/2001` and `44` is parsed as `01/01/2044`.
To be honest I am also confusing why `23` and `24` are not parsed but it is definitely related to how `Date` parses them.

**Comment by bzz** [comment]
Looks great to me and CI is green.

Thank you @Peilin-Yang for prompt responce and @felizbear for reviews!

Merging to master, if there is no further discussion.

**Pull request closed without merging**